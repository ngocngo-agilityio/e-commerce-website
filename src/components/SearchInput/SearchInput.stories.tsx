import type { Meta, StoryObj } from '@storybook/react';

import SearchInput from '.';

const meta = {
  title: 'SearchInput',
  component: SearchInput,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof SearchInput>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    value: '',
    onChange: () => {},
  },
};

export const HaveValue: Story = {
  args: {
    value: 'abcd',
    onChange: () => {},
  },
};
