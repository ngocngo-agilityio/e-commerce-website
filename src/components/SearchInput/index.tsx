// Libs
import { ChangeEvent, memo } from 'react';
import {
  Input,
  InputGroup,
  InputProps,
  InputRightElement,
} from '@chakra-ui/react';
import { useDebounceCallback } from 'usehooks-ts';

// Assets
import { SearchIcon } from '@assets';

interface Props extends InputProps {
  defaultValue?: string;
  placeholder?: string;
  onChange: (e: ChangeEvent<HTMLInputElement>) => void;
}

const SearchInput = ({
  defaultValue,
  placeholder = 'Search Here....',
  onChange,
  ...rest
}: Props): JSX.Element => {
  const handleOnChange = useDebounceCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      onChange(e);
    },
    500,
  );

  return (
    <InputGroup>
      <Input
        aria-label="Product Search"
        placeholder={placeholder}
        variant="outline"
        size="base"
        borderRadius="base"
        pr="80px"
        defaultValue={defaultValue}
        onChange={handleOnChange}
        {...rest}
      />
      <InputRightElement
        w="56px"
        h="24px"
        my="12px"
        borderLeft="1px"
        borderColor="outlineInput.borderColor"
      >
        <SearchIcon />
      </InputRightElement>
    </InputGroup>
  );
};

export default memo(SearchInput);
