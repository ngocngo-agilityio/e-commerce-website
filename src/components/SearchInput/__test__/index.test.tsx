// Libs
import { fireEvent, render, screen } from '@testing-library/react';

// Components
import SearchInput from '..';

const mockProps = {
  value: 'abc',
  onChange: jest.fn(),
};

jest.useFakeTimers();

describe('SearchInput component', () => {
  test('should render SearchInput successfully', () => {
    const { container } = render(<SearchInput {...mockProps} />);

    expect(container).toBeInTheDocument();
  });

  test('should match snapshot for SearchInput', () => {
    const { container } = render(<SearchInput {...mockProps} />);

    expect(container).toMatchSnapshot();
  });

  test('should call onChange when typing value', () => {
    render(<SearchInput {...mockProps} />);

    const searchInput = screen.getByPlaceholderText('Search Here....');

    fireEvent.change(searchInput, { target: { value: 'test' } });

    jest.advanceTimersByTime(1000);

    expect(mockProps.onChange).toHaveBeenCalled();
  });
});
