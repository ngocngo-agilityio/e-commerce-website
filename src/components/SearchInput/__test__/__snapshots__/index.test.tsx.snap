// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput component should match snapshot for SearchInput 1`] = `
<div>
  <div
    class="chakra-input__group css-bx0blc"
    data-group="true"
  >
    <input
      aria-label="Product Search"
      class="chakra-input css-1ms85xl"
      placeholder="Search Here...."
      value="abc"
    />
    <div
      class="chakra-input__right-element css-1gz66tq"
    >
      <svg
        class="chakra-icon css-g7h3az"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          clip-path="url(#clip0_590_644)"
        >
          <path
            d="M11 2C15.968 2 20 6.032 20 11C20 15.968 15.968 20 11 20C6.032 20 2 15.968 2 11C2 6.032 6.032 2 11 2ZM11 18C14.867 18 18 14.867 18 11C18 7.132 14.867 4 11 4C7.132 4 4 7.132 4 11C4 14.867 7.132 18 11 18ZM19.485 18.071L22.314 20.899L20.899 22.314L18.071 19.485L19.485 18.071Z"
            fill="#FF5B19"
          />
        </g>
        <clippath
          id="clip0_590_644"
        >
          <rect
            fill="white"
            height="24"
            width="24"
          />
        </clippath>
      </svg>
    </div>
  </div>
</div>
`;
